File: backend/schemas/summary.py
"""Pydantic-схемы для модуля «Генерация конспектов»."""

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


# ---------------------------------------------------------------------------
# Netology structure (programs / modules / webinars)
# ---------------------------------------------------------------------------

class NetologyProgramResponse(BaseModel):
    """Программа / профессия."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    netology_id: str
    title: str
    program_type: str


class NetologyModuleResponse(BaseModel):
    """Модуль / дисциплина."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    netology_id: str
    title: str
    program_id: UUID


class NetologyLessonItemResponse(BaseModel):
    """Вебинар / видео для выбора."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    netology_id: str
    title: str
    item_type: str
    has_vtt: bool


# ---------------------------------------------------------------------------
# Conspect content (structured LLM output)
# ---------------------------------------------------------------------------

class DefinitionItem(BaseModel):
    """Определение термина."""

    term: str
    definition: str


class ConspectContent(BaseModel):
    """Структурированное содержимое конспекта (генерируется LLM)."""

    topic: str
    summary: str
    key_points: list[str]
    definitions: list[DefinitionItem]
    difficulty: int = Field(ge=1, le=10)


# ---------------------------------------------------------------------------
# Conspect API
# ---------------------------------------------------------------------------

class ConspectCreateRequest(BaseModel):
    """Запрос на генерацию конспекта."""

    lesson_item_id: UUID


class ConspectJobResponse(BaseModel):
    """Статус фоновой задачи."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    status: str
    error_message: str | None
    retry_count: int
    conspect_id: UUID | None
    created_at: datetime
    started_at: datetime | None
    completed_at: datetime | None


class ConspectResponse(BaseModel):
    """Конспект (список)."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    title: str
    topic: str | None
    program_id: UUID | None
    module_id: UUID | None
    lesson_item_id: UUID
    is_edited: bool
    difficulty: int | None
    created_at: datetime
    updated_at: datetime


class ConspectDetailResponse(BaseModel):
    """Полный конспект с содержимым."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    title: str
    topic: str | None
    summary: str | None
    key_points: list[str] | None
    definitions: list[dict[str, Any]] | None
    difficulty: int | None
    raw_vtt_length: int | None = Field(default=None)
    is_edited: bool
    created_at: datetime
    updated_at: datetime


class ConspectUpdateRequest(BaseModel):
    """Редактирование конспекта."""

    title: str | None = None
    summary: str | None = None
    key_points: list[str] | None = None
    definitions: list[DefinitionItem] | None = None


class ConspectListResponse(BaseModel):
    """Список конспектов."""

    items: list[ConspectResponse]
    total: int


# ---------------------------------------------------------------------------
# Search / filters
# ---------------------------------------------------------------------------

class ConspectSearchQuery(BaseModel):
    """Параметры поиска."""

    q: str | None = None
    program_id: UUID | None = None
    module_id: UUID | None = None
    limit: int = Field(default=20, ge=1, le=100)
    offset: int = Field(default=0, ge=0)


# ---------------------------------------------------------------------------
# Knowledge Graph
# ---------------------------------------------------------------------------

class GraphNode(BaseModel):
    """Узел графа знаний."""

    id: str
    label: str
    type: str  # program | module | conspect
    color: str | None = None


class GraphEdge(BaseModel):
    """Ребро графа знаний."""

    source: str
    target: str


class KnowledgeGraphResponse(BaseModel):
    """Данные для графа знаний."""

    nodes: list[GraphNode]
    edges: list[GraphEdge]

File: frontend/src/api/summary.ts
/**
 * API-клиент для модуля «Генерация конспектов».
 */

const API_BASE = "http://localhost:8000/api";

async function fetchWithAuth(path: string, options: RequestInit = {}): Promise<Response> {
  const res = await fetch(`${API_BASE}${path}`, {
    ...options,
    credentials: "include",
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
      ...(options.headers || {}),
    },
  });

  if (!res.ok) {
    const text = await res.text().catch(() => "Unknown error");
    throw new Error(`HTTP ${res.status}: ${text}`);
  }
  return res;
}

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface NetologyProgram {
  id: string;
  netology_id: string;
  title: string;
  program_type: string;
}

export interface NetologyModule {
  id: string;
  netology_id: string;
  title: string;
  program_id: string;
}

export interface NetologyLessonItem {
  id: string;
  netology_id: string;
  title: string;
  item_type: string;
  has_vtt: boolean;
}

export interface ConspectJob {
  id: string;
  status: string;
  error_message: string | null;
  retry_count: number;
  conspect_id: string | null;
  created_at: string;
  started_at: string | null;
  completed_at: string | null;
}

export interface DefinitionItem {
  term: string;
  definition: string;
}

export interface Conspect {
  id: string;
  title: string;
  topic: string | null;
  summary: string | null;
  key_points: string[] | null;
  definitions: DefinitionItem[] | null;
  difficulty: number | null;
  raw_vtt_length: number | null;
  is_edited: boolean;
  created_at: string;
  updated_at: string;
}

export interface ConspectListResponse {
  items: Conspect[];
  total: number;
}

// ---------------------------------------------------------------------------
// Programs / Modules / Webinars
// ---------------------------------------------------------------------------

export async function getPrograms(): Promise<NetologyProgram[]> {
  const res = await fetchWithAuth("/summary/programs");
  return res.json();
}

export async function syncPrograms(): Promise<NetologyProgram[]> {
  const res = await fetchWithAuth("/summary/programs/sync", { method: "POST" });
  return res.json();
}

export async function getProgramModules(programId: string): Promise<NetologyModule[]> {
  const res = await fetchWithAuth(`/summary/programs/${programId}/modules`);
  return res.json();
}

export async function getModuleWebinars(moduleId: string): Promise<NetologyLessonItem[]> {
  const res = await fetchWithAuth(`/summary/modules/${moduleId}/webinars`);
  return res.json();
}

// ---------------------------------------------------------------------------
// Conspect Generation
// ---------------------------------------------------------------------------

export async function generateConspect(lessonItemId: string): Promise<ConspectJob> {
  const res = await fetchWithAuth("/summary/conspects/generate", {
    method: "POST",
    body: JSON.stringify({ lesson_item_id: lessonItemId }),
  });
  return res.json();
}

export async function getJobStatus(jobId: string): Promise<ConspectJob> {
  const res = await fetchWithAuth(`/summary/conspects/jobs/${jobId}`);
  return res.json();
}

// ---------------------------------------------------------------------------
// Conspect CRUD
// ---------------------------------------------------------------------------

export async function getConspects(
  q?: string,
  programId?: string,
  moduleId?: string,
  limit = 20,
  offset = 0
): Promise<ConspectListResponse> {
  const params = new URLSearchParams({ limit: String(limit), offset: String(offset) });
  if (q) params.set("q", q);
  if (programId) params.set("program_id", programId);
  if (moduleId) params.set("module_id", moduleId);
  const res = await fetchWithAuth(`/summary/conspects?${params}`);
  return res.json();
}

export async function getConspect(id: string): Promise<Conspect> {
  const res = await fetchWithAuth(`/summary/conspects/${id}`);
  return res.json();
}

export async function updateConspect(
  id: string,
  data: Partial<Pick<Conspect, "title" | "summary" | "key_points" | "definitions">>
): Promise<Conspect> {
  const res = await fetchWithAuth(`/summary/conspects/${id}`, {
    method: "PATCH",
    body: JSON.stringify(data),
  });
  return res.json();
}

export async function deleteConspect(id: string): Promise<void> {
  await fetchWithAuth(`/summary/conspects/${id}`, { method: "DELETE" });
}

export async function getRecentConspects(): Promise<Conspect[]> {
  const res = await fetchWithAuth("/summary/conspects/recent");
  return res.json();
}

// ---------------------------------------------------------------------------
// Knowledge Graph
// ---------------------------------------------------------------------------

export interface GraphNode {
  id: string;
  label: string;
  type: string;
  color: string | null;
}

export interface GraphEdge {
  source: string;
  target: string;
}

export interface KnowledgeGraphData {
  nodes: GraphNode[];
  edges: GraphEdge[];
}

export async function getKnowledgeGraph(): Promise<KnowledgeGraphData> {
  const res = await fetchWithAuth("/summary/knowledge-graph");
  return res.json();
}
